core/src/main/scala/shapeless/generic.scala

```diff
@@ -380,82 +381,85 @@ trait CaseClassMacros extends ReprTypes {
         child.typeSignature // Workaround for <https://issues.scala-lang.org/browse/SI-7755>
-        if (isCaseClassLike(child) || isCaseObjectLike(child))
-          List(child)
-        else if (child.isSealed)
-          collectCtors(child)
-        else
-          abort(s"$child is not case class like or a sealed trait")
+        if (isCaseClassLike(child) || isCaseObjectLike(child)) List(child)
+        else if (child.isSealed) collectCtors(child)
+        else abort(s"$child is not case class like or a sealed trait")
       }
-    }
 
-    if(isProduct(tpe))
+    if (isProduct(tpe)) {
       List(tpe)
-    else if(isCoproduct(tpe)) {
+    } else if (isCoproduct(tpe)) {
       val basePre = prefix(tpe)
       val baseSym = classSym(tpe)
-      val baseTpe =
-        if(!hk) tpe
-        else {
-          val tc = tpe.typeConstructor
-          val paramSym = tc.typeParams.head
-          val paramTpe = paramSym.asType.toType
-          appliedType(tc, paramTpe)
-        }
+      val baseTpe = if (!hk) tpe else {
+        val tc = tpe.typeConstructor
+        appliedType(tc, tc.typeParams.head.asType.toType)
+      }
+
       val baseArgs = baseTpe.dealias.typeArgs
+      def ctorOf(sym: ClassSymbol) = {
+        import c.internal._
 
-      val ctorSyms = collectCtors(baseSym).sortBy(_.fullName)
-      val ctors =
-        ctorSyms flatMap { sym =>
-          def normalizeTermName(name: Name): TermName =
-            TermName(name.toString.stripSuffix(" "))
-
-          def substituteArgs: List[Type] = {
-            val subst = c.internal.thisType(sym).baseType(baseSym).typeArgs
-            sym.typeParams.map { param =>
-              val paramTpe = param.asType.toType
-              baseArgs(subst.indexWhere(_.typeSymbol == paramTpe.typeSymbol))
-            }
+        def substituteArgs = {
+          val subst = thisType(sym).baseType(baseSym).typeArgs
+          for (param <- sym.typeParams)
+            yield baseArgs(subst.indexWhere(_.typeSymbol == param))
+        }
+
+        // Handle anonymous subclasses.
+        val isNamed = !isAnonOrRefinement(sym)
+        if (isNamed && sym.isStatic) {
+          if (sym.isModuleClass) sym.toType
+          else appliedType(sym.toType, substituteArgs)
+        } else {
+          // Look for a path from the macro call site to the subtype.
+          val owner = sym.owner
+          val owners = ownerChain(if (isNamed) owner else owner.owner)
+          val prePaths = for (pre <- Iterator.iterate(basePre)(prefix).takeWhile(_ != NoPrefix))
+            yield (pre, owners.iterator.dropWhile(pre.baseType(_) == NoType))
+
+          // Find a path from a (sub-)prefix or the enclosing owner.
+          val (pre0, path) = prePaths.find(_._2.nonEmpty).getOrElse {
+            val enclosing = ownerChain(enclosingOwner)
+            val common = owners zip enclosing indexWhere { case (o1, o2) => o1 != o2 }
+            (NoPrefix, if (common < 0) Iterator.empty else owners.iterator drop common - 1)
```

Comment:
This is a last attempt. If none of prefixes worked or there was no prefix, maybe there is a path from the enclosing owner of the macro call site to the subclass. Note that there is always a common owner due to `_root_`. If all owners are common, we just reference the raw type, otherwise we use the differing owners as a path (we drop the common to accommodate nesting and relative accessibility).